File: core/ReverseLinkedListII/ReverseLinkedListII.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def reverseBetween(head: ListNode, left: int, right: int) -> ListNode:
    pre_head = ListNode()
    pre_head.next = head
    res, count = pre_head, 1
    while count < left:
        pre_head = pre_head.next
        count += 1

    left_node = pre_head.next
    cur_node, next_node = left_node, left_node.next
    while count < right:
        temp = next_node.next
        next_node.next = cur_node
        cur_node = next_node
        next_node = temp
        count += 1

    pre_head.next = cur_node
    left_node.next = next_node

    return res.next

File: core/ReverseLinkedListII/ReverseLinkedListII.java
package com.dbc;

public class ReverseLinkedListII {
    public class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode preHead = new ListNode();
        preHead.next = head;
        ListNode res = preHead;
        int count = 1;
        while (count < left) {
            preHead = preHead.next;
            count++;
        }
        ListNode leftNode = preHead.next;
        ListNode curNode = leftNode, nextNode = leftNode.next;
        while (count < right) {
            ListNode temp = nextNode.next;
            nextNode.next = curNode;
            curNode = nextNode;
            nextNode = temp;
            count++;
        }
        preHead.next = curNode;
        leftNode.next = nextNode;

        return res.next;
    }
}
